File: wlr/OGUR/OGUR/Strategies/TargetSet.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Xna.Framework;
using SPX.Entities;
using OGUR.Creatures;
using OGUR.Entities;
using SPX.Core;

namespace OGUR.Strategies
{
    public class TargetSet
    {
        private List<ICreature> m_targets = new List<ICreature>();
        private List<int> m_targetTypes = new List<int>(); 
        private readonly ICreature m_parent;

        public TargetSet(ICreature parent)
        {
            m_parent = parent;
        }

        public ICreature AddTarget(ICreature target)
        {
            m_targets.Add(target);
            return target;
        }

        public void AddTargetTypes(params int[] creatureTypes)
        {
            foreach(var type in creatureTypes)
            {
                m_targetTypes.Add(type);
            }
        }

        public void AddTargets(ICreature source)
        {
            m_targets.AddRange(source.GetTargets().m_targets);
            m_targetTypes.AddRange(source.GetTargets().m_targetTypes);
        }

        public IEnumerable<ICreature> AddTargets(IEnumerable<ICreature> targets)
        {
            if (targets != null)
            {
                m_targets.AddRange(targets);
            }
            return targets;
        }


        private List<ICreature> m_calculatedTargets;
        private float dist;
        ICreature closest;
        public ICreature FindClosest()
        {
            closest = null;
            var closestDistance = float.MaxValue;
            if (m_targets != null)
            {
                foreach (var target in m_targets)
                {
                    dist = Point2.DistanceSquared(target.GetLocation(), m_parent.GetLocation());
                    if (dist < closestDistance)
                    {
                        closest = target;
                        closestDistance = dist;
                    }
                }
                foreach(var creatureType in m_targetTypes)
                {
                    m_calculatedTargets = EntityManager.GetActors(creatureType).Select(a=>a as ICreature).ToList();
                    foreach(var creature in m_calculatedTargets)
                    {
                        if (creature != m_parent)
                        {
                            dist = Point2.DistanceSquared(creature.GetLocation(), m_parent.GetLocation());
                            if (dist < closestDistance)
                            {
                                closest = creature;
                                closestDistance = dist;
                            }
                        }
                    }
                }
            }
            return closest;
        }

        public IEntity GetCollidedTarget(Entity source)
        {
            foreach(var target in m_targets)
            {
                if(target.Contains(source.GetLocation()))
                {
                    return target;
                }
            }

            foreach (var creatureType in m_targetTypes)
            {
                foreach (var target in EntityManager.GetActorsAt(source.GetLocation()))
                {
                    if (target.ActorType()==creatureType || 
                        (creatureType == CreatureType.NONPLAYER && target.ActorType()!=CreatureType.PLAYER) || 
                        (creatureType == CreatureType.PLAYER && target.ActorType()==CreatureType.PLAYER))
                    {
                        return target as IEntity;
                    }
                }
            }
            return null;
        }
    }
}

File: proj/src/aigilas/strategies/TargetSet.java
package aigilas.strategies;

import aigilas.creatures.BaseCreature;
import sps.bridge.ActorType;
import sps.core.Point2;
import sps.entities.Entity;
import sps.entities.EntityManager;
import sps.entities.IActor;
import sps.entities.IEntity;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TargetSet

{
    private final List<BaseCreature> _targets = new ArrayList<>();
    private final List<ActorType> _targetActorTypes = new ArrayList<>();
    private final List<Integer> _targetEntityTypes = new ArrayList<>();
    private final BaseCreature _parent;

    public TargetSet(BaseCreature parent)

    {
        _parent = parent;
    }

    public BaseCreature addTarget(BaseCreature target)

    {
        _targets.add(target);
        return target;
    }

    public void addTargetTypes(ActorType... actorTypes) {
        Collections.addAll(_targetActorTypes, actorTypes);
    }

    public void addEntityTypes(int... entityTypes) {
        for (int type : entityTypes) {
            _targetEntityTypes.add(type);
        }
    }

    public void addTargets(BaseCreature source)

    {
        _targets.addAll(source.getTargets()._targets);
        _targetActorTypes.addAll(source.getTargets()._targetActorTypes);
    }

    public List<BaseCreature> addTargets(List<BaseCreature> targets)

    {
        if (targets != null) {
            _targets.addAll(targets);
        }
        return targets;
    }

    private List<BaseCreature> _calculatedTargets;
    private float dist;
    BaseCreature closest;

    public BaseCreature findClosest() {
        closest = null;
        float closestDistance = Float.MAX_VALUE;
        if (_targets != null) {
            for (BaseCreature target : _targets) {
                dist = Point2.distanceSquared(target.getLocation(), _parent.getLocation());
                if (dist < closestDistance) {
                    closest = target;
                    closestDistance = dist;
                }
            }
            for (ActorType actorType : _targetActorTypes) {
                _calculatedTargets = new ArrayList<>();
                for (IEntity entity : EntityManager.get().getActors(actorType)) {
                    _calculatedTargets.add((BaseCreature) entity);
                }

                for (BaseCreature creature : _calculatedTargets) {
                    if (creature != _parent) {
                        dist = Point2.distanceSquared(creature.getLocation(), _parent.getLocation());
                        if (dist < closestDistance) {
                            closest = creature;
                            closestDistance = dist;
                        }
                    }
                }
            }
        }
        return closest;
    }

    public IEntity getCollidedTarget(Entity source)

    {
        for (BaseCreature target : _targets) {
            if (target.contains(source.getLocation())) {
                return target;
            }
        }

        for (ActorType actorType : _targetActorTypes) {
            for (IActor target : EntityManager.get().getActorsAt(source.getLocation())) {
                if (target.getActorType() == actorType || (actorType == ActorType.NONPLAYER && target.getActorType() != ActorType.PLAYER) || (actorType == ActorType.PLAYER && target.getActorType() == ActorType.PLAYER)) {
                    return target;
                }
            }
        }
        return null;
    }

    public List<ActorType> getTargetActorTypes() {
        return _targetActorTypes;
    }
}
